### ISSUE TITLE
Client certificate auth not working with allow_anonymous=false

### ISSUE THREAD

**Issue by cocowalla** [emqx/emqx#1966]
#### Environment

- OS: Alpine Linux
- Erlang/OTP: 21.0.7
- EMQ: v3.0-beta.3

#### Description

I'm trying to use client certificates for authentication. I'm using stock configuration, with the following changes (I'm running EMQ in a Docker container, hence the use of environment variables to change config):

````
EMQX_ALLOW_ANONYMOUS=false
EMQX_LISTENER__SSL__EXTERNAL__CACERTFILE=/opt/emqx/etc/certs/cacert.pem
EMQX_LISTENER__SSL__EXTERNAL__VERIFY=verify_peer
EMQX_LISTENER__SSL__EXTERNAL__FAIL_IF_NO_PEER_CERT=true
EMQX_LISTENER__SSL__EXTERNAL__PEER_CERT_AS_USERNAME=cn
````

When clients try to connect, it fails and I see this in the EMQ logs:

`emqx_1.1.12918h6kykt0@linuxkit-00155d3ad131    | (emqx@10.0.22.3)1> 2018-11-16 22:22:11.643 [error] MyClientId@10.255.0.2:41186 [MQTT] Username 'undefined' login failed for auth_modules_not_found`

Strangely enough, it works if I remove `EMQX_ALLOW_ANONYMOUS=false` - clients connect, and I see the correct username for the connections in the dashboard. Stranger still is that it works even if the user doesn't exist in EMQ!

Any idea how I can get clients to authenticate using their client certificate (with CN as a username), without using any plugin authentication modules?

**Comment by terry-xiaoyu** [comment]
Ha, not strange.
```
## Value: cn | en | crt
listener.ssl.external.peer_cert_as_username = cn
```
This only let you use cn as username of your clients, but you still need a plugin to validate the username in auth plugins.

EMQ doesn't maintain any client information, your databases such as pgsql does. 

If the only thing you'd like to do is validating the client certificate, set EMQX_ALLOW_ANONYMOUS=true

**Comment by cocowalla** [comment]
Ah, I assumed with no auth plugins it would validate against the list of users created from the dashboard/CLI (that's RabbitMQ's behaviour), but now it makes sense :smile:

Is this behaviour available as a plugin (just validating user exists)?

**Comment by terry-xiaoyu** [comment]
Yes, use https://github.com/emqx/emqx-auth-username.

**Comment by cocowalla** [comment]
@terry-xiaoyu I've tried that, and now I get:

`emqx_1.1.xid7xqi4pyda@linuxkit-00155d3ad131    | (emqx@10.0.23.3)1> 2018-11-17 19:38:02.275 [error] MyClientId@10.255.0.2:41260 [MQTT] Username 'undefined' login failed for password_undefined`

I still have `peer_cert_as_username = cn` set, so I'm not sure why it shows `Username 'undefined'`.

Here's what I did:

1. Added `emqx_auth_username` to the list of plugins in `EMQX_LOADED_PLUGINS`
2. Verified that `emqx_auth_username` is mentioned in the logs when the broker starts
3. Verified in the dashboard that the `emqx_auth_username` plugin has started
4. In the dashboard, I added a user with the same name as the client's certificate CN

In that last step it did force me to enter a password for the user, even though I don't need one - I presume this is ignored when `peer_cert_as_username = cn`?
